### ISSUE TITLE
[RFC][Pipelining] RNG state communication, avoid seed checkpoint

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#139304]
### Background

During model initialization, each PP stage initializes their own portion of the model.  This is preferrable over initializing the model all in one place and then transferring it to each stage, due to the high memory requirements of initializing an entire large model in one place.

Ignoring random states, each PP stage would start from the same RNG seed and each model chunk could have similar or even identical weights.  This would likely lead to numerical issues during training.  It also makes loss-curves of PP training less comparable with loss-curves of non-PP training, which lowers confidence in PP training and adds obstacles to debugging. 

Currently, torchtitan implements a workaround where the whole model is initialized on CPU offline in one process, saved to disk, and then loaded by each pipeline stage.  (Loading does not require loading the whole model into ram, only the portion used locally).  This results in consistent model initialization but at a runtime and UX cost of generating and using the seed checkpoint.

### Proposal
Instead, we can manage the RNG state automatically inside torch.pipelining, to avoid identical weights.

**Option 1. Use different seeds per stage**

We could simply re-seed the RNG on each PP stage, without worrying about actual initialization parity with non-PP training.

This option is potentially faster since it allows each stage to initialize in parallel.  It should also gaurantee 'safe' initializations, since each stage would have no chance of duplicate or similar init weights.  However, this initialization would not 'match' the initialization values for non-PP training, leading to less comparability between loss curves.

**Option 2. Sequentially initialize**

After stage 0 initializes, it could summarize its own RNG states and send them to stage 1, which in turn loads the states and initializes, then saves new states and sends them to stage 2, etc.  This results in serializing initialization, which would lead to an observable slowdown at startup time, but also matches perfectly (bitwise) with the initialization values that would be observed in a non-PP training.

This option requires more complexity:
- virtual-stages means more than one round of communication across PP ranks is needed
- V-schedules means sometimes, neighboring stages exist on the same rank and thus seed communication should be skipped for those cases

Note: seed save/restore can be implemented like this: https://gist.github.com/wconstab/6bdea055eff9a7904fdae595c6cdac6e

It might be worth implementing both options and experimenting with the tradeoffs.  I suspect most users would prefer option 1 if its significantly faster, and we can prove it leads to equally accurate training.

**Comment by wconstab** [comment]
One API idea:

```
-            m.init_weights(buffer_device=buffer_device)
+            with torch.pipelining.set_stage_rng(mode=args.experimental.pipeline_init_mode):
+                # mode = 'parallel (default)'
+                # __enter__: configure a new RNG seed that is unique per stage
+                # __exit__: NoOp
+                #
+                # mode = 'serial'
+                # __enter__: 'recv' RNG states and configure them
+                # __exit__, 'send' latest RNG states to next stage
+                m.init_weights(buffer_device=buffer_device)
```

Issue: need to know (a) current stage idx (to determine whether to send/recv),
and (b) overall schedule layout (interleave, V, etc), to know which rank to send to for next stage

**Comment by wconstab** [comment]
```
with schedule.set_rng_for_init(m, mode):
    m.init_weights(buffer_device=buffer_device)
```

schedule obj already knows the mapping of stage to rank and which rank has the 'next' stage, so that solves part of the issue above.

passing 'current stage_idx' into the context manager is also an option, but i wonder if passing model chunk (m) in is just as good, because it could be looked up by matching

**Comment by wconstab** [comment]
`schedule.apply_init_fn(lambda m: m.init_weights(buffer_device=buffer_device))`

we can actually move the whole 'for loop' inside the schedule.  This way schedule takes care of iterating over its local model chunks and knowing their stage_idx, as well as handling communication as needed.

**Comment by wconstab** [comment]
> Preserving the loss curve comparability is a nice property of option 2 as people put a lot of time verifying the accuracy. I'm wondering if slowness is really a problem. It would be N x slower than option1 where N is the stage number.

yes, it would be N times slower where N is typically 8-32 or maybe 64 (including virtual stages not just pipeline ranks)

note that preserving matching loss is not entirely achieved by option 2, but if the model does not use randomness during training then it is.

I would agree to implement it.  I am leaning towards implementing both modes and letting it be a flag.

**Comment by tianyu-l** [comment]
The reasoning sounds good and the having the two options is great.

Some random thoughts on removing the N-time slowness:

I know that we have decoupled PP split and `init_weights`, in that `init_weights` no longer knows the whole model. But for now let's assume `init_weights` still knows the whole model, and knows what's in this PP rank and what's not, then can we do:
1. For layers/modules **not on** the current PP rank, init with FakeTensor (or whatever way we can proceed CUDA RNG states as if for real tensors);
2. For layers/modules **on** the current PP rank, init as usual.

This way we avoid the communication and the N-time slowness caused by serialization.

If we can't do 1, we can alternatively do:
1. For layers/modules **not on** the current PP rank, init as usual but then delete it immediately;
2. For layers/modules **on** the current PP rank, init as usual.

This way we still have the N-time slowness but avoid the communication and serialization. For super large model parts, it may cause high peak memory unnecessary on the current rank.

**Comment by wconstab** [comment]
> For layers/modules not on the current PP rank, init with FakeTensor (or whatever way we can proceed CUDA RNG states as if for real tensors);

I am pretty sure this is not possible.  FakeTensor execution is really just 'meta device' execution.  (A FakeTensor is tensor subclass that dispatches to the 'meta' device and also tracks more metadata about which 'device' it belong on).  Meta kernels basically just do shape propagation, but they don't call any actual cuda kernels.  The RNG state for cuda would only be touched by the code that launches the cuda kernels, iiuc.  In case i'm wrong, i'll look into this a bit more.

> For layers/modules not on the current PP rank, init as usual but then delete it immediately;

This could definitely work.   if initialization slowness is an issue, this approach would also be somewhat slow.  It would cut out the P2P communication delays (N-1 of them), but it would still serialize N layers of init.  We should do some experiments about how much time is spent in layer init vs p2p comms.  


regarding both of the suggestions- we would have to change the way the pp code partitions the model.  right now it partitions first (on fake model), then materializes empty, then initialized.  The changed version would have to move the 'materialize empty' step somewhere inside the init function, so that it 'knows' to keep parts of the model on fake device, or, to only move one layer to device at a time, initialize it, delete it, etc.  In other words, it complicates the model init function a bit.

**Comment by fegin** [comment]
FakeTensor is used to do shape propagation. I agree with @wconstab , FakeKernel won't change RNG state. Even if it does, the RNG state will not be the correct one.

**Comment by tianyu-l** [comment]
Got it, thanks for explaining! Then what you suggested sounds optimal to me.

**Comment by kwen2501** [comment]
> this initialization would not 'match' the initialization values for non-PP training, leading to less comparability between loss curves

Is equality between loss curves is a goal or a non-goal? 
I tend to think different training method would by default have different behaviors. Users may accept the difference if there is important gain in other aspects, like scalability, performance, etc.
